packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,385 @@
+    // NOTE: for API level 16 the metrics will exclude window decor.
+    DisplayMetrics metrics = new DisplayMetrics();
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
+      windowManager.getDefaultDisplay().getRealMetrics(metrics);
+    } else {
+      windowManager.getDefaultDisplay().getMetrics(metrics);
+    }
+
+    // Calculate physical size.
+    double widthInches = metrics.widthPixels / (double) metrics.xdpi;
+    double heightInches = metrics.heightPixels / (double) metrics.ydpi;
+    double diagonalSizeInches = Math.sqrt(Math.pow(widthInches, 2) + Math.pow(heightInches, 2));
+
+    if (diagonalSizeInches >= 3.0 && diagonalSizeInches <= 6.9) {
+      // Devices in a sane range for phones are considered to be Handsets.
+      return DeviceType.HANDSET;
+    } else if (diagonalSizeInches > 6.9 && diagonalSizeInches <= 18.0) {
+      // Devices larger than handset and in a sane range for tablets are tablets.
+      return DeviceType.TABLET;
+    } else {
+      // Otherwise, we don't know what device type we're on/
+      return DeviceType.UNKNOWN;
+    }
+  }
+
+  private String getSerial() {
+    if (android.os.Build.VERSION.SDK_INT < 26) {
+      return Build.SERIAL;
+    } else {
+      try {
+        int permissionCheck = ContextCompat.checkSelfPermission(mContext, Manifest.permission.READ_PHONE_STATE);
+
+        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
+          ActivityCompat.requestPermissions(mActivity, new String[]{Manifest.permission.READ_PHONE_STATE}, REQUEST_READ_PHONE_STATE);
+
+        } else {
+          return Build.getSerial();
+        }
+      } catch (SecurityException se) {
+        Log.e(TAG, se.getMessage());
+      }
+      return null;
+    }
+  }
+
+  private boolean isTablet() {
+    return deviceType.getValue() == "Tablet";
+  }
+
+  @ExpoMethod
+  public void getFreeDiskStorageAsync(Promise promise) {
+    try {
+      BigInteger storage = this.getFreeDiskStorage();
+      String stringValue = storage.toString();
+      promise.resolve(stringValue);
+    } catch (NullPointerException e) {
+      e.printStackTrace();
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getIPAddressAsync(Promise promise) {
+    try {
+      Integer ipAddress = getWifiInfo().getIpAddress();
+      // Convert little-endian to big-endianif needed
+      if (ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
+        ipAddress = Integer.reverseBytes(ipAddress);
+      }
+
+      byte[] ipByteArray = BigInteger.valueOf(ipAddress).toByteArray();
+
+      String ipAddressString;
+      try {
+        ipAddressString = InetAddress.getByAddress(ipByteArray).getHostAddress();
+      } catch (Exception ex) {
+        Log.e("WIFIIP", "Unable to get host address.");
+        ipAddressString = null;
+      }
+      promise.resolve(ipAddressString);
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
```

Comment:
If we print stacks then we should find a unified way to do it

Reply:
Sure. I just removed the printStackTrace and decided to only log in Android console and promise rejection.